MAINT: bump min NumPy to 1.25.2, min Python to 3.11

tylerjereddy opened the pull request:
* Bump the minimum supported NumPy version to ~`1.24.4`~ `1.25.2` ~ahead of the upcoming SciPy `1.15.0` release~. See discussion related to SPEC 0/community support standards at https://github.com/scipy/scipy/pull/21982#discussion_r1873996863. ~Normally I'd prefer to do this a little farther out from branching, but should probably respect the developer desire to modernize progressively over time as well.~

* It is likely that there are some other subtle opportunities to simplify (i.e., testing shims no longer needed, etc.), but most of those kinds of things are less critical anyway and just "nice to have/find."

TODO/known issues:

- [x] Linux ARM CI job still needs a shim (don't think it got checked/adjusted on my fork). EDIT: taken care of by gh-22446
- [x] bumping Python and/or NumPy version seems to be causing `test_stats.py::TestRegression::test_regressZEROX` to fail in a wide variety of CI scenarios
- [x] ~`test_mpmath.py::TestSystematic::test_besselj_complex` fails after version bump in some CI scenarios~ this is gh-22395
- [x] there are two tricky Linux CI jobs that folks should review with skepticism--`test_venv_install` and `distro_multiple_pythons` both jump through some hoops to test tricky things, and while they are passing on my fork, it is not hard to imagine that a small mistake could lead to a loss of the "spirit" of what they are trying to test
- [x] `pixi` lockfile updating may need to happen eventually, I believe `git grep` found a bunch of `3.10` strings in there

tylerjereddy committed:
MAINT: bump min NumPy to 1.24.4

* Bump the minimum supported NumPy version to
`1.24.4` ahead of the upcoming SciPy `1.15.0` release.

rgommers reviewed (commented):
There's no benefit to doing this now, we've already done all the work to keep 1.23.5 working. There are potential downsides for users (smaller support window), and a very small amount of risk to do it the same weekend as a branch. So I'd recommend merging this only after the `1.15.x` branch creation.

At that point it looks like we can at once also drop Python 3.10 and NumPy 1.24

tylerjereddy commented:
ok, I bumped the milestone

rgommers commented:
> At that point it looks like we can at once also drop Python 3.10 and NumPy 1.24

@tylerjereddy do you want to do that in this PR? I can take care of it if you prefer, but only after the holidays.

tylerjereddy commented:
> @tylerjereddy do you want to do that in this PR? I can take care of it if you prefer, but only after the holidays.

Sure, I can give it a try.

tylerjereddy committed:
WIP, MAINT: PR 22012 revisions

* Bump the minimum numpy version to `1.25.2`, based
on reviewer feedback.

* Initial work to bump minimum Python version to `3.11`,
based on reviewer feedback.

[ci skip]

tylerjereddy commented:
I pushed in a commit to bump to NumPy `1.25.2` instead of `1.24.4`. Still working on dropping Python `3.10.x` though, so I'll circle back to this between release duties.

rgommers commented:
Thanks Tyler. It's not that urgent, so no worries.

rgommers commented:
Python 3.10 aarch64 wheel build jobs are giving me a headache for some weird reason. Rather than worry about that, I'd like to get this in. Do you have time to resume this @tylerjereddy, or can I push to your branch here?

tylerjereddy commented:
Let's see what I can get done today, IIRC I may need to work on my fork for a bit to sustinably bump/fix some of the jobs.

tylerjereddy committed:
MAINT: PR 22012 revisions

* More adjustments to handle removal of support for Python `3.10.`

tylerjereddy committed:
MAINT: PR 22012 revisions

* For the Linux Python debug CI job, bump the base Ubuntu version
now that we don't support Python 3.10. The new Python debug
version in use for that job is Python 3.12.3.

* For the Linux `test_venv_install` CI job, bump the base Ubuntu
version now that we don't support Python 3.10.

* For the Linux `distro_multiple_pythons` CI job, also bump
the base Ubuntu version so that we use Python 3.11 (non-default)
and Python 3.12 (default).

tylerjereddy commented:
@rgommers I've summarized the current status observed on my fork (https://github.com/tylerjereddy/scipy/pull/81) in the updated TODO list in original PR comment above.

I cherry-picked the relevant commits into this PR as well, so hopefully the list of issues matches reasonably well.

rgommers commented on .github/workflows/linux.yml:
```diff
@@ -422,8 +422,10 @@ jobs:
-          # `python3-dev` yields Python 3.10 on Ubuntu 22.04
-          sudo apt install -y python3-dev python3.11-dev ninja-build pkg-config libatlas-base-dev liblapack-dev
+          sudo apt install software-properties-common
+          sudo add-apt-repository ppa:deadsnakes/ppa
```
This isn't desirable, `deadsnakes` isn't as stable as the main Ubuntu package repositories.

It should be fine to keep using `python3-dev`, that installs 3.12 in Ubuntu 24.04. The job doesn't need to run on the lowest supported Python version, the only point is to check a distro-supplied Python interpreter.

rgommers replied:
I'll check locally before pushing an update

rgommers replied:
Ah I see the problem, it doesn't have to be the lowest supported version, but it should be a non-default version. And Ubuntu 24.04 doesn't seem to have any. `noble-updates` may gain `3.13` at some point, but until then maybe this is the best option (at least easier than switching to a different distro).

rgommers committed:
Merge branch 'main' into treddy_bump_numpy

rgommers commented:
Checked off most todo's, just `test_regressZEROX` left (assuming I didn't mess up the merge from `main`).

tylerjereddy commented:
Thanks, yeah I was a little nervous about some of the changes, but didn't have better ideas. Let me know if you want me to check on that last test, I'm just working on something else at the moment but can circle back tomorrow if you want.

tylerjereddy reviewed (commented):
From a little more grepping while avoiding other stuff this evening, these simplifications are also possible, but could probably be dealt with a bit later:

- `test_gh_17992` no longer needs a warning suppression for NumPy < `1.24`
- Two tests in `scipy/sparse/tests/test_base.py` have comments explaining improvements that can be made when `1.24` support is dropped

tylerjereddy commented:
The cross-listed issue contains a small patch that fixes the `test_regressZEROX` problem--I pinged Matt to confirm before we pull that over or just patch it separately...

tylerjereddy committed:
Merge branch 'main' into treddy_bump_numpy

tylerjereddy commented:
Ok, the `test_regressZEROX` matter was resolved on `main` and I've updated this branch accordingly--let's see if the CI is finally happy here.

tylerjereddy commented:
CI is passing, so I think this PR is ready for a round of review.

Once merged, I think we will (as usual) discover all kinds of places where NumPy version-gated shims can be simplified, as noted for a few cases I already saw above (but I bet there are way more).

rgommers reviewed (approved):
Excellent, LGTM. Thanks Tyler!
